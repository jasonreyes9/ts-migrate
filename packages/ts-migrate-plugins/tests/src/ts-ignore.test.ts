import tsIgnorePlugin from '../../src/plugins/ts-ignore';
import { mockPluginParams, mockDiagnostic } from '../test-utils';

describe('ts-ignore plugin', () => {
  it('adds ignore comment', async () => {
    const text = "comsole.log('Hello');";
    const result = await tsIgnorePlugin.run(
      mockPluginParams({
        text,
        semanticDiagnostics: [mockDiagnostic(text, 'comsole')],
      }),
    );
    expect(result).toMatchSnapshot();
  });

  it('custom comment', async () => {
    const text = "comsole.log('Hello');";
    const result = await tsIgnorePlugin.run(
      mockPluginParams({
        text,
        semanticDiagnostics: [mockDiagnostic(text, 'comsole')],
        options: {
          messagePrefix: 'custom message prefix',
        },
      }),
    );
    expect(result).toMatchSnapshot();
  });

  it('adds ignore comment with ts-ignore', async () => {
    const text = "comsole.log('Hello');";
    const result = await tsIgnorePlugin.run(
      mockPluginParams({
        text,
        semanticDiagnostics: [mockDiagnostic(text, 'comsole')],
        options: { useTsIgnore: true },
      }),
    );
    expect(result).toMatchSnapshot();
  });

  it('adds ignore comment in jsx', async () => {
    const text = `import React from 'react';

function Foo() {
  return (
    <div>
      <DoesNotExist />
    </div>
  );
}

export default Foo;
`;

    const result = await tsIgnorePlugin.run(
      mockPluginParams({
        fileName: 'Foo.tsx',
        text,
        semanticDiagnostics: [mockDiagnostic(text, 'DoesNotExist')],
      }),
    );

    expect(result).toMatchSnapshot();
  });
  it('adds ignore comment in jsx with Fragment', async () => {
    const text = `import React from 'react';

function Foo() {
  return (
    <>
      <DoesNotExist />
    </>
  );
}

export default Foo;
`;

    const result = await tsIgnorePlugin.run(
      mockPluginParams({
        fileName: 'Foo.tsx',
        text,
        semanticDiagnostics: [mockDiagnostic(text, 'DoesNotExist')],
      }),
    );

    expect(result).toMatchSnapshot();
  });

  it('truncates error message if too long', async () => {
    const text = "comsole.log('Hello');";
    const result = await tsIgnorePlugin.run(
      mockPluginParams({
        text,
        semanticDiagnostics: [
          mockDiagnostic(text, 'comsole', {
            messageText: 'This message is too long to print and should be truncated',
          }),
        ],
      }),
    );
    expect(result).toMatchSnapshot();
  });

  it('use message limit option to avoid error message truncation', async () => {
    const text = "comsole.log('Hello');";
    const result = await tsIgnorePlugin.run(
      mockPluginParams({
        text,
        semanticDiagnostics: [
          mockDiagnostic(text, 'comsole', {
            messageText:
              'This message is long, but should not be translated because of the messageLimit option value',
          }),
        ],
        options: { messageLimit: 100 },
      }),
    );
    expect(result).toMatchSnapshot();
  });

  it('use message limit option to truncate a error message', async () => {
    const text = "comsole.log('Hello');";
    const result = await tsIgnorePlugin.run(
      mockPluginParams({
        text,
        semanticDiagnostics: [
          mockDiagnostic(text, 'comsole', {
            messageText:
              'This message is too long, and should be truncated because of the messageLimit option value',
          }),
        ],
        options: { messageLimit: 75 },
      }),
    );
    expect(result).toMatchSnapshot();
  });

  it('does not add ignore comment for webpackChunkName', async () => {
    const text = `const getComponent = normalizeLoader(() =>
  import(
    /* webpackChunkName: "Component_async" */
    './this_module_does_not_exist'
  ),
);
`;

    const result = await tsIgnorePlugin.run(
      mockPluginParams({
        text,
        semanticDiagnostics: [mockDiagnostic(text, 'this_module_does_not_exist')],
      }),
    );

    expect(result).toMatchSnapshot();
  });

  it('handles error within ternary when true', async () => {
    const text = `function foo() {
  return something
    ? doesNotExist
    : other;
}
`;

    const result = await tsIgnorePlugin.run(
      mockPluginParams({
        text,
        semanticDiagnostics: [mockDiagnostic(text, 'doesNotExist')],
      }),
    );

    expect(result).toMatchSnapshot();
  });

  it('handles error within ternary when false', async () => {
    const text = `function foo() {
  return something
    ? other
    : doesNotExist;
}
`;

    const result = await tsIgnorePlugin.run(
      mockPluginParams({
        text,
        semanticDiagnostics: [mockDiagnostic(text, 'doesNotExist')],
      }),
    );

    expect(result).toMatchSnapshot();
  });

  it('handles error within ternary jsx expression', async () => {
    const text = `function Foo() {
  return someBoolean
    ? <ComponentA />
    : <ComponentB />;
}
`;

    const result = await tsIgnorePlugin.run(
      mockPluginParams({
        fileName: 'Foo.tsx',
        text,
        semanticDiagnostics: [mockDiagnostic(text, 'ComponentA')],
      }),
    );

    expect(result).toMatchSnapshot();
  });

  it('handles error within ternary property access', async () => {
    const text = `function Foo() {
  return someBoolean
    ? this.props.doesNotExist
    : <SomeComponent />;
}
`;

    const result = await tsIgnorePlugin.run(
      mockPluginParams({
        fileName: 'Foo.tsx',
        text,
        semanticDiagnostics: [mockDiagnostic(text, 'doesNotExist')],
      }),
    );

    expect(result).toMatchSnapshot();
  });

  it('handles neighboring eslint disable comment', async () => {
    const text = `function foo() {
  // eslint-disable-next-line
  return doesNotExist;
}
`;

    const result = await tsIgnorePlugin.run(
      mockPluginParams({
        text,
        semanticDiagnostics: [mockDiagnostic(text, 'doesNotExist')],
      }),
    );

    expect(result).toMatchSnapshot();
  });

  it('handles multiline ternary', async () => {
    const text = `function Foo() {
  return someBoolean ? (
    <ComponentA
      doesNotExist="fail"
    />
  ) : (
    <ComponentB />
  );
}
`;

    const result = await tsIgnorePlugin.run(
      mockPluginParams({
        fileName: 'Foo.tsx',
        text,
        semanticDiagnostics: [mockDiagnostic(text, 'doesNotExist')],
      }),
    );

    expect(result).toMatchSnapshot();
  });

  it('handles single line ternary', async () => {
    const text = `function Foo() {
  return someBoolean ? <ComponentA /> : <ComponentB />;
}
`;

    const result = await tsIgnorePlugin.run(
      mockPluginParams({
        fileName: 'Foo.tsx',
        text,
        semanticDiagnostics: [mockDiagnostic(text, 'ComponentA')],
      }),
    );

    expect(result).toMatchSnapshot();
  });
});
