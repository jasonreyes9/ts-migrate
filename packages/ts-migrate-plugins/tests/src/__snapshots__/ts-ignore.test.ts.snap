// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ts-ignore plugin adds ignore comment 1`] = `
"// @ts-expect-error ts-migrate(123) FIXME: diagnostic message
comsole.log('Hello');"
`;

exports[`ts-ignore plugin adds ignore comment in jsx 1`] = `
"import React from 'react';

function Foo() {
  return (
    <div>
      {/* @ts-expect-error ts-migrate(123) FIXME: diagnostic message */}
      <DoesNotExist />
    </div>
  );
}

export default Foo;
"
`;

exports[`ts-ignore plugin adds ignore comment in jsx with Fragment 1`] = `
"import React from 'react';

function Foo() {
  return (
    <>
      {/* @ts-expect-error ts-migrate(123) FIXME: diagnostic message */}
      <DoesNotExist />
    </>
  );
}

export default Foo;
"
`;

exports[`ts-ignore plugin adds ignore comment with ts-ignore 1`] = `
"// @ts-ignore ts-migrate(123) FIXME: diagnostic message
comsole.log('Hello');"
`;

exports[`ts-ignore plugin does not add ignore comment for webpackChunkName 1`] = `
"const getComponent = normalizeLoader(() =>
  import(
    /* webpackChunkName: \\"Component_async\\" */
    './this_module_does_not_exist'
  ),
);
"
`;

exports[`ts-ignore plugin handles error within ternary jsx expression 1`] = `
"function Foo() {
  return someBoolean
    ? // @ts-expect-error ts-migrate(123) FIXME: diagnostic message
      <ComponentA />
    : <ComponentB />;
}
"
`;

exports[`ts-ignore plugin handles error within ternary property access 1`] = `
"function Foo() {
  return someBoolean
    ? // @ts-expect-error ts-migrate(123) FIXME: diagnostic message
      this.props.doesNotExist
    : <SomeComponent />;
}
"
`;

exports[`ts-ignore plugin handles error within ternary when false 1`] = `
"function foo() {
  return something
    ? other
    : // @ts-expect-error ts-migrate(123) FIXME: diagnostic message
      doesNotExist;
}
"
`;

exports[`ts-ignore plugin handles error within ternary when true 1`] = `
"function foo() {
  return something
    ? // @ts-expect-error ts-migrate(123) FIXME: diagnostic message
      doesNotExist
    : other;
}
"
`;

exports[`ts-ignore plugin handles multiline ternary 1`] = `
"function Foo() {
  return someBoolean ? (
    <ComponentA
      // @ts-expect-error ts-migrate(123) FIXME: diagnostic message
      doesNotExist=\\"fail\\"
    />
  ) : (
    <ComponentB />
  );
}
"
`;

exports[`ts-ignore plugin handles neighboring eslint disable comment 1`] = `
"function foo() {
  // @ts-expect-error ts-migrate(123) FIXME: diagnostic message
  // eslint-disable-next-line
  return doesNotExist;
}
"
`;

exports[`ts-ignore plugin handles single line ternary 1`] = `
"function Foo() {
  // @ts-expect-error ts-migrate(123) FIXME: diagnostic message
  return someBoolean ? <ComponentA /> : <ComponentB />;
}
"
`;

exports[`ts-ignore plugin truncates error message if too long 1`] = `
"// @ts-expect-error ts-migrate(123) FIXME: This message is too long to print and should be tr... Remove this comment to see the full error message
comsole.log('Hello');"
`;

exports[`ts-ignore plugin use message limit option to avoid error message truncation 1`] = `
"// @ts-expect-error ts-migrate(123) FIXME: This message is long, but should not be translated because of the messageLimit option value
comsole.log('Hello');"
`;

exports[`ts-ignore plugin use message limit option to truncate a error message 1`] = `
"// @ts-expect-error ts-migrate(123) FIXME: This message is too long, and should be truncated because of the messageLim... Remove this comment to see the full error message
comsole.log('Hello');"
`;
